### PR TITLE
Force file sync after writing file via container in test

### PR DESCRIPTION
**Pull request by davidz627**
/kind failing-test

This fix might fix flakes in the migration test suite. Force sync of written file after writing

/assign @msau42 @misterikkit 
/cc @ddebroy

```release-note
NONE
```

### CODE REVIEW

**Comment by davidz627** [comment]
/hold
I didn't notice the outer loop. fixing now

**Comment by davidz627** [comment]
/hold cancel
fixed to collate a single final error for each file which is used to fail the test after all files checked

File: test/e2e/storage/pd.go [hunk 1]
```diff
@@ -451,29 +453,34 @@ func countReadyNodes(c clientset.Interface, hostName types.NodeName) int {
...
-				break
+				err = fmt.Errorf("read content <%q> does not match expected content <%q>", v, expectedContents)
+				continue
 			}
+			// All checks completed - Success
+			success = true
+			break
+		}
+		if !success {
+			errs = append(errs, fmt.Errorf("could not verify that contents of file %v : %v", filePath, err))
```
Comment (ddebroy) [inline test/e2e/storage/pd.go]: nit: `verify that contents of` => `verify contents of`

File: test/e2e/storage/pd.go [hunk 2]
```diff
@@ -451,29 +453,34 @@ func countReadyNodes(c clientset.Interface, hostName types.NodeName) int {
...
 		}
-		framework.ExpectEqual(strings.TrimSpace(value), strings.TrimSpace(expectedContents))
```
Comment (ddebroy) [inline test/e2e/storage/pd.go]: The above check was redundant and therefore being completely removed, correct?

**Review by ddebroy — COMMENTED** [review]
Just one nit. Otherwise lgtm.

File: test/e2e/storage/pd.go (hunk 2 above)
Comment (davidz627) [reply test/e2e/storage/pd.go]: yes

File: test/e2e/storage/pd.go [hunk 3]
```diff
@@ -451,29 +453,33 @@ func countReadyNodes(c clientset.Interface, hostName types.NodeName) int {
 }
 
 func verifyPDContentsViaContainer(f *framework.Framework, podName, containerName string, fileAndContentToVerify map[string]string) {
+	var err error
+	var v string
```
Comment (msau42) [inline test/e2e/storage/pd.go]: does v have to be at the top? it's not actually used outside the loop

File: test/e2e/storage/pd.go [hunk 4]
```diff
@@ -451,29 +453,33 @@ func countReadyNodes(c clientset.Interface, hostName types.NodeName) int {
...
+	errs := []error{}
 	for filePath, expectedContents := range fileAndContentToVerify {
-		var value string
-		// Add a retry to avoid temporal failure in reading the content
-		for i := 0; i < maxReadRetry; i++ {
-			v, err := f.ReadFileViaContainer(podName, containerName, filePath)
-			value = v
+		// Add a timed retry to avoid temporal failure in reading the content
+		err = nil
+		for start := time.Now(); time.Since(start) < time.Minute; time.Sleep(10 * time.Second) {
```
Comment (msau42) [inline test/e2e/storage/pd.go]: If the tests actually wrote data with O_SYNC, then theoretically we shouldn't even need to retry reading the file?

Comment (davidz627) [reply test/e2e/storage/pd.go]: changed it so that we actually wait for pod 0 to go down first

File: test/e2e/storage/pd.go [hunk 5]
```diff
@@ -451,29 +455,33 @@ func countReadyNodes(c clientset.Interface, hostName types.NodeName) int {
...
 	for filePath, expectedContents := range fileAndContentToVerify {
-		var value string
-		// Add a retry to avoid temporal failure in reading the content
-		for i := 0; i < maxReadRetry; i++ {
-			v, err := f.ReadFileViaContainer(podName, containerName, filePath)
-			value = v
+		var err error
+		err = nil
+		// Add a timed retry to avoid temporal failure in reading the content
+		for start := time.Now(); time.Since(start) < framework.PollShortTimeout; time.Sleep(framework.Poll) {
```
Comment (msau42) [inline test/e2e/storage/pd.go]: Then we can remove retry altogether?

**Comment by davidz627** [comment]
@msau42 lol totally changed this PR. It just forces a sync now

File: test/e2e/storage/pd.go [hunk 6]
```diff
@@ -188,6 +190,8 @@ var _ = utils.SIGDescribe("Pod Disks", func() {
 					ginkgo.By("deleting host0Pod") // delete this pod before creating next pod
 					framework.ExpectNoError(podClient.Delete(host0Pod.Name, podDelOpt), "Failed to delete host0Pod")
 					framework.Logf("deleted host0Pod %q", host0Pod.Name)
+					e2epod.WaitForPodToDisappear(cs, host0Pod.Namespace, host0Pod.Name, labels.Everything(), framework.Poll, framework.PodDeleteTimeout)
```
Comment (msau42) [inline test/e2e/storage/pd.go]: Since we're doing both sync + wait for pod to disappear, then we shouldn't need to retry the read?

Comment (davidz627) [reply test/e2e/storage/pd.go]: yeah I was leaning on the cautious side but I think we should remove this.

File: test/e2e/storage/pd.go [hunk 7]
```diff
@@ -452,28 +455,15 @@ func countReadyNodes(c clientset.Interface, hostName types.NodeName) int {
...
-				size, err := f.CheckFileSizeViaContainer(podName, containerName, filePath)
-				if err != nil {
-					framework.Logf("Error checking file size: %v", err)
-				}
-				framework.Logf("Check file %q size: %q", filePath, size)
-			} else {
-				break
-			}
+		// No retry loop as there should not be temporal based failures
+		v, err := f.ReadFileViaContainer(podName, containerName, filePath)
```
Comment (davidz627) [inline test/e2e/storage/pd.go]: actually what if there is a temporary failure in `kubectl exec` - network issue or something

Comment (msau42) [reply test/e2e/storage/pd.go]: It internally retries: https://github.com/kubernetes/kubernetes/blob/d6a3518ea2ad92fb48e1de6c10f1fbd7e61736c3/test/e2e/framework/framework.go#L496

File: test/e2e/storage/pd.go [hunk 8]
```diff
@@ -452,28 +455,15 @@ func countReadyNodes(c clientset.Interface, hostName types.NodeName) int {
...
+		if err != nil {
```
Comment (msau42) [inline test/e2e/storage/pd.go]: I think you can replace this block with `framework.ExpectNoError`

**Comment by msau42** [comment]
/approve
just one nit

**Comment by davidz627** [comment]
resolved @msau42

**Pull request merged**